spring-integration-file/src/test/java/org/springframework/integration/file/remote/session/ContextHolderRequestHandlerAdviceTests.java

```diff
@@ -0,0 +1,148 @@
+/*
+ * Copyright 2013-2023 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.integration.file.remote.session;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.function.Consumer;
+import java.util.function.Function;
+
+import org.junit.jupiter.api.Test;
+
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.integration.annotation.ServiceActivator;
+import org.springframework.integration.config.EnableIntegration;
+import org.springframework.integration.dsl.MessageChannels;
+import org.springframework.integration.test.util.TestUtils;
+import org.springframework.messaging.Message;
+import org.springframework.messaging.MessageChannel;
+import org.springframework.messaging.PollableChannel;
+import org.springframework.messaging.support.GenericMessage;
+import org.springframework.stereotype.Component;
+import org.springframework.test.context.junit.jupiter.SpringJUnitConfig;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.Mockito.mock;
+
+/**
+ * @author Adel Haidar
+ *
+ * @since 6.1
+ *
+ */
+@SpringJUnitConfig
+public class ContextHolderRequestHandlerAdviceTests {
+
+	@Autowired
+	TestSessionFactory foo;
+
+	@Autowired
+	TestSessionFactory bar;
+
+	@Autowired
+	Config config;
+
+	@Test
+	public void testFlow() {
+		final Map<String, Object> headers = Map.of("test-key", "foo");
+		config.in().send(new GenericMessage<>("any-payload", headers));
+
+		Message<?> received = config.out().receive(0);
```

Comment:
This test is going to pass even without `ContextHolderRequestHandlerAdvice`.
You just really don't demonstrate a usage of this feature in that `TestService`.
Even if that `foo` is selected properly from the `DelegatingSessionFactory`, there is no proof anywhere in the code that we indeed have taken a `foo` `ConnectionFactory`.

Pay attention that `DelegatingSessionFactoryTests` you took ideas from has that `verify(foo.mockSession).list("foo/");` with the meaning that target handler really interacted with mocked `foo` `SessionFactory`.

Your current test does not have that logic.